### ISSUE TITLE
translate cu to cl file

### ISSUE THREAD

**Issue by Gitkingly** [CHIP-SPV/chipStar#568]
Is it able to translate the cuda source file to opencl3.0 host and device files? Do we have any options?

**Comment by pvelesko** [comment]
No translation tools included. Why not run the cuda file directly?

**Comment by Gitkingly** [comment]
Thanks for the fast response. I just want to port some cuda applications to opencl devices which may not have cuda runtime.

**Comment by pvelesko** [comment]
What devices are you interested in? we hope to provide a cuda runtime on as
many as possible.

**Comment by Gitkingly** [comment]
I was told the target is to support all opencl devices, including CPUs/GPUs/FPGAs, not only NVIDIA GPUs. Actually there already is a project did this:
https://github.com/hughperkins/coriander.git.  It is a tool “Build applications written in NVIDIA® CUDA™ code for OpenCL™ 1.2 devices.”

However, it only supports OpenCL1.2 and the project is not maintained since 2 years ago. I am wondering whether this project can do the similar things with OpenCL3.0, it will be very cool if we can.

**Comment by pvelesko** [comment]
Well we do support most CPUs through PoCL. Intel CPUs have an OpenCL runtime that we support as well. Not sure what's the current status on ARM providing an OpenCL 2.1 runtime but it's been a while since I checked. You can also run HIP on Intel, AMD, and Nvidia GPUs.

So it seems that we're only missing FPGAs?

**Comment by pjaaskel** [comment]
@Gitkingly yep that's the idea. chipStar tool differs only a bit from Coriander, we compile device programs to SPIR-V instead of OpenCL 1.2 sources.

**Comment by Gitkingly** [comment]
Thanks @pjaaskel @pvelesko， really appreciate your quick feedbacks to my question. I need some time to dig into the OpenCL and  SPIR-V. -:)
